### PR TITLE
C#: Extract relational patterns

### PR DESCRIPTION
**Pull request by tamasvajk**

### CODE REVIEW

File: csharp/ql/src/semmlecode.csharp.dbscheme [hunk 1]
```diff
@@ -997,11 +997,17 @@ case @expr.kind of
 | 119 = @assign_coalesce_expr
 | 120 = @suppress_nullable_warning_expr
 | 121 = @namespace_access_expr
+/* C# 9.0 */
+| 122 = @lt_pattern
```
Comment (tamasvajk) [inline csharp/ql/src/semmlecode.csharp.dbscheme:1001]: Maybe `lt_pattern_expr` is a better name.

File: csharp/change-notes/2020-11-19-Relational-pattern.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+lgtm,codescanning
+* The `RelationalPatternExpr` and its 4 subclass have been added to support C# 9
```
Comment (hvitved) [inline csharp/change-notes/2020-11-19-Relational-pattern.md:2]: sub classes

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/Patterns/RecursivePattern.cs [hunk 3]
```diff
@@ -6,7 +6,7 @@
 
 namespace Semmle.Extraction.CSharp.Entities.Expressions
 {
-    internal class RecursivePattern : Expression
+    internal partial class RecursivePattern : Expression
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/Patterns/RecursivePattern.cs:9]: Why this change?

File: csharp/ql/src/semmle/code/csharp/exprs/Expr.qll [hunk 4]
```diff
@@ -342,6 +342,45 @@ class ConstantPatternExpr extends PatternExpr {
   override string getAPrimaryQlClass() { result = "ConstantPatternExpr" }
 }
 
+/** A relational pattern, for example `>1` in `x is >1`. */
+abstract class RelationalPatternExpr extends PatternExpr, @relational_pattern_expr {
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/exprs/Expr.qll:346]: Remove `abstract`.

File: csharp/ql/src/semmle/code/csharp/exprs/Expr.qll [hunk 5]
```diff
@@ -342,6 +342,45 @@ class ConstantPatternExpr extends PatternExpr {
...
+  /** Get the operator of this relational pattern. */
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/exprs/Expr.qll:347]: `/** Gets the name of the operator in this pattern. */`

File: csharp/ql/src/semmle/code/csharp/exprs/Expr.qll [hunk 6]
```diff
@@ -342,6 +342,45 @@ class ConstantPatternExpr extends PatternExpr {
...
+  string getOperator() { none() }
+
+  /** Gets the expression of this relational pattern. */
+  Expr getExpr() { result = this.getChild(0) }
+
+  override string toString() { result = getOperator() + " ..." }
+}
+
+/** A 'less than' pattern */
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/exprs/Expr.qll:356]: ``A less-than pattern, for example `< 10` in `x is < 10` .``

File: csharp/ql/src/semmle/code/csharp/exprs/Expr.qll [hunk 7]
```diff
@@ -342,6 +342,45 @@ class ConstantPatternExpr extends PatternExpr {
...
+class LTPattern extends RelationalPatternExpr, @lt_pattern_expr {
+  override string getOperator() { result = "<" }
+
+  override string getAPrimaryQlClass() { result = "LTPattern" }
+}
+
+/** A 'greater than' pattern */
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/exprs/Expr.qll:363]: ``A greater-than pattern, for example `> 10` in `x is > 10` .``

File: csharp/ql/src/semmle/code/csharp/exprs/Expr.qll [hunk 8]
```diff
@@ -342,6 +342,45 @@ class ConstantPatternExpr extends PatternExpr {
...
+class GTPattern extends RelationalPatternExpr, @gt_pattern_expr {
+  override string getOperator() { result = ">" }
+
+  override string getAPrimaryQlClass() { result = "GTPattern" }
+}
+
+/** A 'less than equal' pattern */
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/exprs/Expr.qll:370]: ``A less-than or equals pattern, for example `<= 10` in `x is <= 10` .``

File: csharp/ql/src/semmle/code/csharp/exprs/Expr.qll [hunk 9]
```diff
@@ -342,6 +342,45 @@ class ConstantPatternExpr extends PatternExpr {
...
+class LEPattern extends RelationalPatternExpr, @le_pattern_expr {
+  override string getOperator() { result = "<=" }
+
+  override string getAPrimaryQlClass() { result = "LEPattern" }
+}
+
+/** A 'greater than equal' pattern */
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/exprs/Expr.qll:377]: ``A greater-than or equals pattern, for example `>= 10` in `x is >= 10` .``

**Review by hvitved — CHANGES_REQUESTED** [review]
LGTM. Some trivial comments.

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/Patterns/RecursivePattern.cs (hunk 3 above)
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/Patterns/RecursivePattern.cs:9]: no reason. I removed it.

**Review by hvitved — DISMISSED** [review]

**Review by hvitved — APPROVED** [review]

**Pull request merged**